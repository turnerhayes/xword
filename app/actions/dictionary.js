export const FIND_TERMS = "@XWORD/TERMS/FIND";

export function findTerms(searchArgs) {
	return {
		type: FIND_TERMS,
		payload: searchArgs
	};
	// return (dispatch, getState) => {
	// 	const state = getState();

	// 	if (!state.getIn(["dictionary", "termSearches", searchArgs.pattern])) {
	// 		dispatch({
	// 			type: FIND_TERMS,
	// 			payload: DictionaryUtils.findTerms(searchArgs).then(
	// 				(results) => Map({
	// 					results,
	// 					searchArgs: fromJS(searchArgs),
	// 				})
	// 			),
	// 		});
	// 	}
	// };
}
