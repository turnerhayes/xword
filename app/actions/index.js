export * from "./users";
export * from "./ui";
export * from "./auth";
export * from "./dictionary";
export * from "./puzzle";
