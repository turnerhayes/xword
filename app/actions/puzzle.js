
export const ADD_PUZZLE = "@XWORD/PUZZLES/ADD";

export function addPuzzle({ puzzle, setAsCurrent }) {
	return {
		type: ADD_PUZZLE,
		payload: {
			puzzle,
			setAsCurrent,
		},
	};
}

export const SET_GENERATED_PUZZLE = "@XWORD/PUZZLES/GENERATED/SET";

export function setGeneratedPuzzle({ puzzle }) {
	return {
		type: SET_GENERATED_PUZZLE,
		payload: {
			puzzle
		},
	};
}

export const UPDATE_GENERATED_PUZZLE_CELL = "@XWORD/PUZZLES/GENERATED/UPDATE_CELL";

export function updateGeneratedPuzzleCell({ columnIndex, rowIndex, cell }) {
	return {
		type: UPDATE_GENERATED_PUZZLE_CELL,
		payload: {
			columnIndex,
			rowIndex,
			cell
		},
	};
}

export const UPDATE_GENERATED_PUZZLE_GRID = "@XWORD/PUZZLES/GENERATED/UPDATE_GRID";

export function updateGeneratedPuzzleGrid({ grid }) {
	return {
		type: UPDATE_GENERATED_PUZZLE_GRID,
		payload: {
			grid
		},
	};
}

export const UPDATE_GENERATED_PUZZLE_CLUE = "@XWORD/PUZZLES/GENERATED/UPDATE_CLUE";

export function updateGeneratedPuzzleClue({ clueNumber, clueText, direction }) {
	return {
		type: UPDATE_GENERATED_PUZZLE_CLUE,
		payload: {
			clueNumber,
			clueText,
			direction
		},
	};
}

export const CLEAR_GENERATED_PUZZLE_CLUES = "@XWORD/PUZZLES/GENERATED/CLEAR_CLUES";

export function clearGeneratedPuzzleClues() {
	return {
		type: CLEAR_GENERATED_PUZZLE_CLUES,
	};
}

export const SET_CURRENT_PUZZLE_INDEX = "@XWORD/PUZZLES/SET_CURRENT_INDEX";

export function setCurrentPuzzleIndex({ index }) {
	return {
		type: SET_CURRENT_PUZZLE_INDEX,
		payload: index,
	};
}

export const SET_PUZZLE_CELL_CONTENT = "@XWORD/PUZZLES/SET_CELL_CONTENT";

export function setPuzzleCellContent({ puzzleIndex, position, value }) {
	return {
		type: SET_PUZZLE_CELL_CONTENT,
		payload: {
			puzzleIndex,
			position,
			value,
		},
	};
}
