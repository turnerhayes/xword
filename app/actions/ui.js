export const SET_UI_STATE = "@@XWORD/UI/SET_STATE";

export function setUIState({ section, settings}) {
	return {
		type: SET_UI_STATE,
		payload: {
			section,
			settings,
		},
	};
}
