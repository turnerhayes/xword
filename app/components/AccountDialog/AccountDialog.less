.c_account-dialog {
	.login-link {
		font-size: 2em;

		&,
		&:hover,
		&:active {
			text-decoration: none;
		}
	}

	.provider-icon {
		margin-right: 0.5em;
	}
}
