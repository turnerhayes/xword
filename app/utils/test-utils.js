import React from "react";
import PropTypes from "prop-types";
import configureStore from "redux-mock-store";
import { runSaga as realRunSaga } from "redux-saga";
import { IntlProvider } from "react-intl";
import { mount as enzymeMount, shallow as enzymeShallow } from "enzyme";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import Renderer from "react-test-renderer";

import { translationMessages } from "@app/i18n";
import createReducer from "@app/reducers";

const locale = "en";
const defaultLocale = "en";

const _mockStore = configureStore();

export const mockStore = (initialState, ...args) => {
	if (!initialState) {
		initialState = createReducer()(undefined, {});
	}

	const store = _mockStore(initialState, ...args);

	store.runSaga = () => {};

	return store;
};

const WrapperComponent = ({ children, store }) => (
	<MemoryRouter>
		<Provider store={store}>
			<IntlProvider
				locale={locale}
				defaultLocale={defaultLocale}
				messages={translationMessages[locale]}
			>
				{children}
			</IntlProvider>
		</Provider>
	</MemoryRouter>
);

WrapperComponent.propTypes = {
	children: PropTypes.any,
	store: PropTypes.object,
};

export const mount = (node, {
	store = mockStore(),
} = {}) => enzymeMount(node, {
	wrappingComponent: WrapperComponent,
	wrappingComponentProps: {
		store,
	},
});

export const shallow = (node, {
	store = mockStore(),
} = {}) => enzymeShallow(node, {
	wrappingComponent: WrapperComponent,
	wrappingComponentProps: {
		store
	},
});

export const renderComponent = (node, {
	store = mockStore(),
} = {}) => Renderer.create(
	<WrapperComponent
		store={store}
	>
		{node}
	</WrapperComponent>
);

export async function runSaga({
	state,
	getSaga,
}, ...args) {
	const dispatched = [];

	const rootSaga = await getSaga();

	const dispatchers = [];

	const reducer = createReducer();

	state = state || reducer(undefined, {});

	const getDispatch = (pushToArray = true) => {
		return (action) => {
			if (pushToArray) {
				dispatched.push(action);
			}

			state = reducer(state, action);
		};
	};

	const dispatchWithoutPushing = getDispatch(false);

	const sagaPromise = realRunSaga(
		{
			dispatch: getDispatch(),
			getState: () => state,
			subscribe: (callback) => {
				const wrappedCallback = (action, ...args) => {
					dispatchWithoutPushing(action);
					return callback(action, ...args);
				};

				dispatchers.push(wrappedCallback);

				return () => {
					const index = dispatchers.findIndex(wrappedCallback);

					if (index >= 0) {
						delete dispatchers[index];
					}
				};
			},
		},
		rootSaga,
		...args
	).done;

	return {
		dispatchers,
		dispatched,
		sagaPromise,
	};
}
