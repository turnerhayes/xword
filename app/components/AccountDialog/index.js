import AccountDialog from "./AccountDialog";

export default AccountDialog;
