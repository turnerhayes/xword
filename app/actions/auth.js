import { navigateToUrl } from "@app/utils/navigation";

function getCurrentPagePath() {
	return `${document.location.pathname}${document.location.search}${document.location.hash}`;
}


export const LOGIN = "@@XWORD/AUTH/LOGIN";

export function login({ provider }) {
	const currentPage = getCurrentPagePath();

	if (provider === "facebook") {
		navigateToUrl(`/auth/facebook?redirectTo=${encodeURIComponent(currentPage)}`);
	}
	else if (provider === "google") {
		navigateToUrl(`/auth/google?redirectTo=${encodeURIComponent(currentPage)}`);
	}
	else if (provider === "twitter") {
		navigateToUrl(`/auth/twitter?redirectTo=${encodeURIComponent(currentPage)}`);
	}
	else {
		throw new Error(`Unrecognized login provider "${provider}"`);
	}

	return {
		type: LOGIN,
	};
}

export const LOGOUT = "@@XWORD/AUTH/LOGOUT";

export function logout() {
	const currentPage = getCurrentPagePath();

	navigateToUrl(`/auth/logout?redirectTo=${encodeURIComponent(currentPage)}`);

	return {
		type: LOGOUT,
	};
}
