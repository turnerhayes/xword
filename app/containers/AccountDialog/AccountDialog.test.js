import React from "react";
import { fromJS } from "immutable";
import * as immutableMatchers from "jest-immutable-matchers";
import { MemoryRouter } from "react-router";
import Icon from "@material-ui/core/Icon";
import Button from "@material-ui/core/Button";

import { LOGOUT, LOGIN } from "@app/actions";
import { renderComponent, mount, mockStore } from "@app/utils/test-utils";

import AccountDialog from "./AccountDialog";

const locationProp = Object.getOwnPropertyDescriptor(document, "location");
let locationAssign;

beforeAll(() => {
	jest.addMatchers(immutableMatchers);
});

beforeEach(() => {
	locationAssign = jest.fn();
	locationAssign.mockName("assign");

	Object.defineProperty(document, "location", {
		assign: locationAssign,
	});
});

afterAll(() => {
	Object.defineProperty(window, "location", locationProp);
});

describe("AccountDialog container", () => {
	fit("should pass the logged in user if user is logged in", () => {
		const user = fromJS({
			id: "1",
			isMe: true,
			provider: "facebook",
		});

		const store = mockStore(fromJS({
			users: {
				currentID: user.get("id"),
				items: {
					[user.get("id")]: user,
				},
			},
		}));

		const component = renderComponent(<AccountDialog />, { store, });

		expect(component.toJSON()).toMatchSnapshot();
	});

	it("should pass undefined for the logged in user if user is not logged in", () => {
		const store = mockStore(fromJS({
			users: {
			},
		}));

		const component = renderComponent(<AccountDialog />, { store, });

		expect(component.toJSON()).toMatchSnapshot();
	});

	it("should dispatch a logout action", () => {
		const user = fromJS({
			id: "1",
			isMe: true,
			provider: "facebook",
		});

		const store = mockStore(fromJS({
			users: {
				currentID: user.get("id"),
				items: {
					[user.get("id")]: user,
				},
			},
		}));

		store.dispatch = jest.fn();

		const wrapper = mount(
			(
				<MemoryRouter>
					<AccountDialog
					/>
				</MemoryRouter>
			),
			{
				store,
			},
		);

		const logoutButton = wrapper.find(Button).filterWhere(
			(button) => button.findWhere(
				(el) => el.is(Icon) && el.text() === "log out"
			)
		);

		logoutButton.simulate("click");

		expect(locationAssign).toHaveBeenCalledWith("/auth/logout?redirectTo=blank");

		expect(store.dispatch).toHaveBeenCalledWith({
			type: LOGOUT,
		});
	});

	it("should dispatch a login action", () => {
		const store = mockStore(fromJS({
			users: {
				items: {},
			},
		}));

		store.dispatch = jest.fn();

		const wrapper = mount(
			(
				<MemoryRouter>
					<AccountDialog
						enabledProviders={[ "facebook" ]}
					/>
				</MemoryRouter>
			),
			{
				store,
			},
		);

		const classes = wrapper.find("AccountDialog").prop("classes");

		const loginButton = wrapper.find(`IconButton.${classes.loginLink}`).first();

		loginButton.simulate("click");

		expect(locationAssign).toHaveBeenCalledWith("/auth/facebook?redirectTo=blank");

		expect(store.dispatch).toHaveBeenCalledWith({
			type: LOGIN,
		});
	});
});
