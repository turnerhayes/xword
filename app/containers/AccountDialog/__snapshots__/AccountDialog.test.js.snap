// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountDialog container should pass the logged in user if user is logged in 1`] = `
<div
  className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    className="MuiCardHeader-root"
  >
    <div
      className="MuiCardHeader-content"
    >
      <span
        className="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
      >
        <div>
          <span
            aria-hidden={true}
            className="material-icons MuiIcon-root icon AccountDialog-providerIcon-2"
          >
            facebook
          </span>
          <a
            href="/profile/undefined"
            onClick={[Function]}
          />
        </div>
      </span>
    </div>
  </div>
  <div
    className="MuiCardContent-root"
  />
  <div
    className="MuiCardActions-root MuiCardActions-spacing"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-text"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        <span
          aria-hidden={true}
          className="material-icons MuiIcon-root icon AccountDialog-buttonIcon-3"
        >
          log out
        </span>
        Log out
      </span>
    </button>
  </div>
</div>
`;
